x-pack/plugin/core/src/main/java/org/elasticsearch/snapshots/EncryptedRepository.java

```diff
@@ -0,0 +1,279 @@
+            for (String pathComponent : path) {
+                encryptionMetadataBlobPath = encryptionMetadataBlobPath.add(pathComponent);
+            }
+            return new EncryptedBlobContainerDecorator(this.delegatedBlobStore.blobContainer(path),
+                    this.delegatedBlobStore.blobContainer(encryptionMetadataBlobPath), this.masterSecretKey);
+        }
+    }
+
+    private static class EncryptedBlobContainerDecorator implements BlobContainer {
+
+        private final BlobContainer delegatedBlobContainer;
+        private final BlobContainer encryptionMetadataBlobContainer;
+        private final SecretKey masterSecretKey;
+
+        EncryptedBlobContainerDecorator(BlobContainer delegatedBlobContainer, BlobContainer encryptionMetadataBlobContainer,
+                SecretKey masterSecretKey) {
+            this.delegatedBlobContainer = delegatedBlobContainer;
+            this.encryptionMetadataBlobContainer = encryptionMetadataBlobContainer;
+            this.masterSecretKey = masterSecretKey;
+        }
+
+        @Override
+        public BlobPath path() {
+            return this.delegatedBlobContainer.path();
+        }
+
+        @Override
+        public InputStream readBlob(String blobName) throws IOException {
+            final BytesReference dataDecryptionKeyBytes = Streams.readFully(this.encryptionMetadataBlobContainer.readBlob(blobName));
+            try {
+                SecretKey dataDecryptionKey = unwrapKey(BytesReference.toBytes(dataDecryptionKeyBytes), this.masterSecretKey);
+                Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
+                cipher.init(Cipher.DECRYPT_MODE, dataDecryptionKey, gcmParameterSpec);
+                return new CipherInputStream(this.delegatedBlobContainer.readBlob(blobName), cipher);
+            } catch (InvalidKeyException | NoSuchAlgorithmException | NoSuchPaddingException | InvalidAlgorithmParameterException e) {
+                throw new IOException(e);
+            }
+        }
+
+        @Override
+        public void writeBlob(String blobName, InputStream inputStream, long blobSize, boolean failIfAlreadyExists) throws IOException {
+            try {
+                SecretKey dataEncryptionKey = generateRandomSecretKey();
+                byte[] wrappedDataEncryptionKey = wrapKey(dataEncryptionKey, this.masterSecretKey);
+                try (InputStream stream = new ByteArrayInputStream(wrappedDataEncryptionKey)) {
+                    this.encryptionMetadataBlobContainer.writeBlob(blobName, stream, wrappedDataEncryptionKey.length, failIfAlreadyExists);
+                }
+                Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
+                cipher.init(Cipher.ENCRYPT_MODE, dataEncryptionKey, gcmParameterSpec);
+                this.delegatedBlobContainer.writeBlob(blobName, new CipherInputStream(inputStream, cipher), blobSize, failIfAlreadyExists);
```

Comment:
This seems wrong (and like it'll be a problem potentially!). The `blobSize` for the purposes of cloud stores like S3 must be the exact number of bytes that will be written. This currently works with the `FsRepository` because it doesn't use the size here, but it will write partial (since the encrypted bytes are more than the unencrypted) data with S3 and such.

Can we guess the size up front here in some form (via padding magic or so ... I'm admittedly not that knowledgeable here)?

Reply:
Great catch! We can compute the cyphertext length from the padding and mode we use. It's not complicated. I will make the changes.